### DISCUSSION TITLE
Nodejs CLI to create SVG charts serverside

### DISCUSSION THREAD

**Discussion by harlantwood** [AstroDraw/AstroChart#74] [Ideas]
**Is your feature request related to a problem? Please describe.**

No. 

**Describe the solution you'd like**

I have a [working prototype of generating an SVG on the server side via a new CLI functionality](https://github.com/AstroDraw/AstroChart/compare/main...harlantwood:AstroChart:cli).  It uses `jsdom` to fake a DOM, and then extracts the SVG string, outputting it to stdout, where you can (for example) redirect it to a file.

The current API accepts a few JSON blobs as command line arguments -- it's not really intended to be used directly by humans, but more by other programs.  We could of course later add more granular options which could be more human friendly if desired.

Example of current use:

```sh
bin/astrochart \
  --radix '{"planets":{"Sun":[60],"Moon":[120]},"cusps":[0,30,60,90,120,150,180,210,240,270,300,330]}' \
  --transits '{"planets":{"Mars":[0],"Venus":[180]},"cusps":[0,30,60,90,120,150,180,210,240,270,300,330]}' \
  > chart.svg
```

( note that --settings '{some json}' is not yet in the prototype, but will be soon.

This command generates a standalone SVG that can be opened directly in a browser.  Example SVG generated from the command above:

<img src="https://github.com/AstroDraw/AstroChart/assets/38769/49b3a099-3c22-4aaa-bb97-baa039044a2e" width="400">

----

There are multiple options for how this feature could be integrated (or not) here:

1. Directly into this repo.  
    - The current prototype uses `tsm` to execute typescript directly, which is useful only if you clone the repo
    - If we wanted to go npx (think `npx astrochart --options ...`) we would probably move to nodejs using the webpack transpiled output
1. In a separate repo that uses this one as a dependency.
    - This one is interesting if we want to keep this repo as purely client side, but still want the CLI functionality to exist elsewhere.
1. Not at all.  If this is not of general interest, I'll just use it in my project, which would be fine 😄

**Comment by afucher** [comment]
hey @harlantwood this is a really good idea! 💡 

About the approach, I prefer the 2nd. 
If you want, I can create a new repo in the organization and add you there. Just give me a name.
Or I think that you can create it in your account, and once you think is ready, can transfer to the organization if you want.

WDYT?

**Reply by harlantwood** [reply]
Sure, go ahead and create it in the org… how about AstroChartCLI?

one thing to think about: the npm package name is how you would invoke it via npx — so if/when we create an npm package, would be cool to have a short awesome name.  Since astrochart is taken, maybe astrodraw, if available…

**Reply by erenkulaksiz** [reply]
-- and please dont use webpack -- you can just use tsc-watch

![image](https://github.com/AstroDraw/AstroChart/assets/44942737/eec8d7d0-c845-4586-9643-bcddd8be2f4f)

**Reply by afucher** [reply]
@harlantwood about the name, I don't know if since it is from a org if the name will be @astrodraw/<package name> in the npx command. If yes, we can go to a simpler name.. WDYT?

**Reply by afucher** [reply]
some list names that came into my mind:
 - astrochart-cli
 - astro-cli
 - chart-cli
 - astrodraw-cli

**Reply by harlantwood** [reply]
Good point, user will have to do

npx @astrodraw/something

so how about we make it as ergo as possible:

npx @astrodraw/cli

— my second choice is chart-cli

**Reply by afucher** [reply]
I made some tests but seems that `npx` does not take into consideration the organization 😢

**Reply by harlantwood** [reply]
Hmm just checking that you’re trying this with an npm org scope, not a github org?

**Reply by harlantwood** [reply]
FWIW this SO seems to point to it working with npm org scopes:

https://stackoverflow.com/a/58649605

**Reply by afucher** [reply]
yeah, I have the astroraw org in npm, if you check the astrochart package is @astrodraw/astrochart
but for npx it does not accept, at least I tried some ways and also look for docs or any other package that has this, and did not found

**Reply by afucher** [reply]
> FWIW this SO seems to point to it working with npm org scopes:
> 
> https://stackoverflow.com/a/58649605

I can try this, but there is a comment in the answer saying that it also does not worked 🤷

**Reply by harlantwood** [reply]
Hmm maybe you did this, but the npx enabling code would probably have to be checked into the main branch to work.  Can you link to your commits or PR?

**Comment by harlantwood** [comment]
FYI due to other things heating up, not sure I’ll be able to land this anytime soon, but will try!

**Reply by roabhi** [reply]
This is awesome. Being able to render chart server side opens a whole new world of possibilities in combination con swisseph. Any updates on the implementation of the feature?

**Comment by harlantwood** [comment]
Update: not expecting to have time to do further work on this, but for anyone interested, feel free to fork my prototype, which works great:

https://github.com/AstroDraw/AstroChart/compare/main...harlantwood:AstroChart:cli

cc @roabhi @afucher

**Reply by roabhi** [reply]
Hey dumb question. How do I fork your prototype?. I may have the time to work on this atm

**Reply by harlantwood** [reply]
Nice. Just head over to https://github.com/harlantwood/AstroChart/tree/cli - clone that repo or add a remote to it to your local repo - this work is on the ‘cli’ branch.
